warehouse/retry.go

```diff
@@ -0,0 +1,197 @@
+			updated_at = NOW()
+		WHERE %[1]s`,
+		clausesQuery,
+	)
+	retryReq.API.log.Info(preparedStatement)
+
+	// Executing the statement
+	result, err := retryReq.API.dbHandle.Exec(preparedStatement, clausesArgs...)
+	if err != nil {
+		return
+	}
+
+	// Getting rows affected
+	rowsAffected, err = result.RowsAffected()
+	return
+}
+
+func (retryReq *RetryRequest) whereClauses(sourceIDs []string) (clauses []string, clausesArgs []interface{}) {
+	// SourceID
+	if retryReq.SourceID != "" {
+		clauses = append(clauses, fmt.Sprintf(`source_id = %s`, retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, retryReq.SourceID)
+	} else {
+		clauses = append(clauses, fmt.Sprintf(`source_id = ANY(%s)`, retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, pq.Array(sourceIDs))
+	}
+
+	// DestinationID
+	if retryReq.DestinationID != "" {
+		clauses = append(clauses, fmt.Sprintf(`destination_id = %s`, retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, retryReq.DestinationID)
+	}
+
+	// DestinationType
+	if retryReq.DestinationType != "" {
+		clauses = append(clauses, fmt.Sprintf(`destination_type = %s`, retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, retryReq.DestinationType)
+	}
+
+	// Status deciding based on forcefully retrying
+	if !retryReq.ForceRetry {
+		clauses = append(clauses, fmt.Sprintf("status = %s", retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, Aborted)
+	}
+
+	// UploadIDs or Retry Interval
+	if len(retryReq.UploadIds) != 0 {
+		clauses = append(clauses, fmt.Sprintf("id = ANY(%s)", retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, pq.Array(retryReq.UploadIds))
+	} else {
+		clauses = append(clauses, fmt.Sprintf("created_at > NOW() - %s * INTERVAL '1 HOUR'", retryQueryPlaceHolder))
+		clausesArgs = append(clausesArgs, retryReq.GetRetryInterval())
+	}
+	return
+}
+
+func (retryReq *RetryRequest) successMessage(uploadsRetried int64) string {
+	if uploadsRetried == 0 {
+		return "No retried uploads to sync for this destination"
+	}
+	return fmt.Sprintf("Retried successfully %d syncs", uploadsRetried)
+}
+
+func (retryReq *RetryRequest) validateReq() (err error) {
+	if !retryReq.API.enabled || retryReq.API.log == nil || retryReq.API.dbHandle == nil {
+		err = errors.New("warehouse service is not initialized")
+		return
+	}
+
+	if retryReq.SourceID == "" && retryReq.DestinationID == "" && retryReq.WorkspaceID == "" {
+		retryReq.API.log.Errorf(`WH: Please provide valid request parameters while retrying jobs for workspaceId: %s, sourceId: %s, destinationId: %s`,
+			retryReq.WorkspaceID,
+			retryReq.SourceID,
+			retryReq.DestinationID,
+		)
+		err = errors.New("please provide valid request parameters while retrying jobs with workspaceId or sourceId or destinationId")
+		return
+	}
+	return
+}
+
+func (retryReq *RetryRequest) GetRetryInterval() int64 {
+	if retryReq.IntervalInHours <= 0 {
+		return config.GetInt64("Warehouse.retryIntervalInHrs", 12)
+	}
+	return retryReq.IntervalInHours
```

Comment:
Can we improve this logic a bit as it feels like Line 196 will never be executed ?

Reply:
Everytime line 196 will be called.